src/models/StyledComponent.js

```diff
@@ -117,52 +169,72 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
     }
 
     componentWillUnmount() {
-      if (this.unsubscribe) {
-        this.unsubscribe()
+      const { componentStyle } = this.constructor
+      if (componentStyle.isStaticRules) {
+        return
       }
+
+      this.unsubscribeFromContext()
     }
 
-    render() {
-      const { innerRef } = this.props
+    makeClassName(): string {
+      // I know this is ugly, but I have a really surprigins JSPerf
+      // pen showing this is ~11x faster than running [...classes].filter(Boolean).join(' ')
+      // (and uses way less memory from the two array allocations in ^)
+      // and  ~ 36 % faster than always adding a whitespace and then calling .trim at the end
+      const { className: propsClassName } = this.props
+      const { className: attrsClassName } = this.attrs
       const { generatedClassName } = this.state
-      const { styledComponentId, target } = this.constructor
+      const { styledComponentId } = this.constructor
 
-      const isTargetTag = isTag(target)
+      let className = propsClassName || ''
+      if (styledComponentId) {
+        if (className.length) {
+          className += ' '
+        }
+        className += styledComponentId
+      }
+      if (attrsClassName) {
+        if (className.length) {
+          className += ' '
+        }
+        className += attrsClassName
+      }
+      if (generatedClassName) {
+        if (className.length) {
+          className += ' '
+        }
+        className += generatedClassName
+      }
+
+      return className
+    }
 
-      const className = [
-        this.props.className,
-        styledComponentId,
-        this.attrs.className,
-        generatedClassName,
-      ].filter(Boolean).join(' ')
+    render() {
+      // $FlowFixMe
+      const { innerRef } = this.props
+      const { target } = this.constructor
 
-      const baseProps = {
+
+      const propsForElement = {
         ...this.attrs,
-        className,
+        ...this.props,
```

Comment:
This introduces a regression and removes the prop filtering: `(!isTargetTag || validAttr(propName))`.

Please revert it :)